### PR TITLE
Fixes for <exclude/> support, with tests and docs.

### PR DESCRIPTION
**Pull request by davedawkins**
Documentation on rebasing links
Issues #786 , #785

### CODE REVIEW

File: FSharp.Formatting.sln
Comment (nojaf) [inline FSharp.Formatting.sln:86]: `TestLib3` isn't showing up under the right folder for me:
image

File: docs/apidocs.fsx
Comment (nojaf) [inline docs/apidocs.fsx:183]: Should this not be `<exclude />`? `</exclude>` without opening tag doesn't strike me as valid XML.

File: src/FSharp.Formatting.ApiDocs/GenerateHtml.fs
Comment (nojaf) [inline src/FSharp.Formatting.ApiDocs/GenerateHtml.fs:710]: Should we consider having the filtering inside `.EntityInfos` instead?
Based on the usage of this member I think this makes sense.

File: src/FSharp.Formatting.ApiDocs/GenerateModel.fs
Comment (nojaf) [inline src/FSharp.Formatting.ApiDocs/GenerateModel.fs:2277]: These last two lines of the comment can be omitted?

File: src/FSharp.Formatting.ApiDocs/GenerateModel.fs
Comment (nojaf) [inline src/FSharp.Formatting.ApiDocs/GenerateModel.fs:2286]: This journal entry seems a tad excessive. You explained the reasoning not ten lines ago.

File: src/FSharp.Formatting.ApiDocs/GenerateModel.fs-debug
Comment (nojaf) [inline src/FSharp.Formatting.ApiDocs/GenerateModel.fs-debug:1]: Should this file exist?

**Review by nojaf — CHANGES_REQUESTED** [review]
This looks like a solid PR. I have some minor nitpicks along the way but I think we can take this in.
At the same time, I'm no expert so @nhirschey do you want to glance over this one as well?

@davedawkins if you update the `RELEASE_NOTES.md` file with your fix and a new version (say `17.2.1`) a new version would be released once we merged this to main.
If you don't need a new version that's fine as well.

File: src/FSharp.Formatting.ApiDocs/GenerateModel.fs-debug
Comment (davedawkins) [reply src/FSharp.Formatting.ApiDocs/GenerateModel.fs-debug:1]: No - will fix

File: src/FSharp.Formatting.ApiDocs/GenerateModel.fs
Comment (davedawkins) [reply src/FSharp.Formatting.ApiDocs/GenerateModel.fs:2286]: Well, it's a single line pointing at the explanation, in response to the question raised in the previous developer's comment "why are we here". How do you want to improve this?

File: src/FSharp.Formatting.ApiDocs/GenerateHtml.fs
Comment (davedawkins) [reply src/FSharp.Formatting.ApiDocs/GenerateHtml.fs:710]: I will take a look at EntifyInfos. My preference was to do it in the logic that calls `checkAccess`, but I believe that's before we've processed the comments.

File: docs/apidocs.fsx
Comment (davedawkins) [reply docs/apidocs.fsx:183]: Yes, a typo. Sorry :-/

File: FSharp.Formatting.sln
Comment (davedawkins) [reply FSharp.Formatting.sln:86]: I will fix this.

File: src/FSharp.Formatting.ApiDocs/GenerateModel.fs
Comment (davedawkins) [reply src/FSharp.Formatting.ApiDocs/GenerateModel.fs:2286]: I removed all the comments. I see your point (to an extent), however: the original comment either needs a response (in context), or just removing. My approach is not to remove comments, but respond to them in a timeline to help out developers down the line.  I think however nothing is lost - if you think it's excessive, others may do too.

**Comment by davedawkins** [comment]
All updates made as requested.

File: src/FSharp.Formatting.ApiDocs/GenerateModel.fs
Comment (nojaf) [inline src/FSharp.Formatting.ApiDocs/GenerateModel.fs:2263]: The match is looking a bit funny now.
Perhaps go with:
```fsharp
            // sum can be null with null/empty el.Value when an non-"<summary>" XML element appears
            // as the only '///' documentation command:
            //
            // 1.
            //  // Not triple-slash ccomment
            //  /// <exclude/>
            //
            // 2.
            //  /// <exclude/>
            if isNull sum then
                let doc, nsels = readXmlCommentAsHtmlAux false ctx.UrlMap el cmds

                let nsdocs = readNamespaceDocs ctx.UrlMap nsels
                cmds, doc, nsdocs
            elif ctx.MarkdownComments then
                readMarkdownCommentAndCommands ctx sum.Value el cmds
            else
                if sum.Value.Contains("<exclude") then
                    cmds.["exclude"] <- ""

                    printfn
                        "Warning: detected \"<exclude/>\" in text of \"<summary>\" for \"%s\". Please see https://fsprojects.github.io/FSharp.Formatting/apidocs.html#Classic-XML-Doc-Comments"
                        xmlSig

                readXmlCommentAndCommands ctx sum.Value el cmds
```
instead.

**Review by nojaf — APPROVED** [review]
Looks good to me.

File: src/FSharp.Formatting.ApiDocs/GenerateModel.fs
Comment (davedawkins) [reply src/FSharp.Formatting.ApiDocs/GenerateModel.fs]: Agreed, your fix applied

**Comment by nhirschey** [comment]
Thanks so much @davedawkins. @nojaf sure, I should be able take a look in a few hours.

**Comment by davedawkins** [comment]
Please hold off from applying this PR. I think I've found an issue with excluding `type`.

**Comment by davedawkins** [comment]
There's no issue with excluding type as such, but the module will still list an excluded type. Additionally, the search index is listing excluded items, so I'm going to fix both those issues today. Unless requested otherwise, I'm going to check those changes into this branch (and PR).

**Comment by nojaf** [comment]
Hello @davedawkins, thank you so much for your persistence here!

I think having the index changes in this PR is ok. I took a peek at the latest changes and everything is still ok for me.

**Comment by davedawkins** [comment]
@nojaf, no problem. I'm determined to get Sutil's API documented! Thank you for help too.

**Review by nhirschey — APPROVED** [review]

**Comment by nhirschey** [comment]
Wonderful set of fixes @davedawkins. Thank you! It looks good to me as well. 

I went ahead and merged the PR to trigger the release. It should hit the feed soon.

**Pull request merged**
